llama.cpp

```diff
@@ -4547,6 +4389,12 @@ struct llm_build_context {
                     cb(Vcur, "Vcur", il);
                 }
 
+                // these nodes are added to the graph together so that they are not reordered
+                // by doing so, the number of splits in the graph is reduced
+                ggml_build_forward_expand(gf, Qcur);
+                ggml_build_forward_expand(gf, Kcur);
+                ggml_build_forward_expand(gf, Vcur);
```

Comment:
More generally, I've been thinking that grouping the weights for K, Q, and V together would be beneficial in terms of performance. Currently the same hidden state is converted to FP16 three times for each of the matrices. Also smaller ggml graphs mean less kernel launch overhead and larger CUDA grids suffer less from tail effects.